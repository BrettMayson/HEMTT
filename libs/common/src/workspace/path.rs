use std::{hash::Hasher, io::Write, sync::Arc};

use tracing::trace;
use vfs::VfsPath;

use super::{Error, Workspace};

#[allow(clippy::module_name_repetitions)]
#[derive(Clone, PartialEq, Eq)]
/// A path in a workspace
pub struct WorkspacePath {
    pub(crate) path: VfsPath,
    pub(crate) workspace: Arc<Workspace>,
}

impl WorkspacePath {
    #[must_use]
    /// Returns the underlying [`VfsPath`]
    pub const fn vfs(&self) -> &VfsPath {
        &self.path
    }

    /// join a path to the workspace path
    ///
    /// # Errors
    /// [`Error::Vfs`] if the path could not be joined
    pub fn join(&self, path: impl AsRef<str>) -> Result<Self, Error> {
        let path = self.path.join(path)?;
        Ok(Self {
            path,
            workspace: self.workspace.clone(),
        })
    }

    /// Create a file in the workspace
    ///
    /// # Errors
    /// [`Error::Vfs`] if the file could not be created
    pub fn create_file(&self) -> Result<Box<dyn Write + Send>, Error> {
        self.path.create_file().map_err(Into::into)
    }

    /// Create a directory in the workspace
    ///
    /// # Errors
    /// [`Error::Vfs`] if the directory could not be created
    pub fn create_dir(&self) -> Result<(), Error> {
        self.path.create_dir()?;
        Ok(())
    }

    /// Check if the path exists
    ///
    /// # Errors
    /// [`Error::Vfs`] if the path could not be checked
    pub fn exists(&self) -> Result<bool, Error> {
        self.path.exists().map_err(Into::into)
    }

    /// Check if the path is a file
    ///
    /// # Errors
    /// [`Error::Vfs`] if the path could not be checked
    pub fn is_file(&self) -> Result<bool, Error> {
        self.path.is_file().map_err(Into::into)
    }

    /// Check if the path is a directory
    ///
    /// # Errors
    /// [`Error::Vfs`] if the path could not be checked
    pub fn is_dir(&self) -> Result<bool, Error> {
        self.path.is_dir().map_err(Into::into)
    }

    /// Read the path to a string
    ///
    /// # Errors
    /// [`Error::Vfs`] if the path could not be read
    pub fn read_to_string(&self) -> Result<String, Error> {
        self.path
            .read_to_string()
            .map(|s| s.replace('\r', ""))
            .map_err(Into::into)
    }

    /// Open the file for reading
    ///
    /// # Errors
    /// [`Error::Vfs`] if the file could not be opened
    pub fn open_file(&self) -> Result<Box<dyn vfs::SeekAndRead + Send>, Error> {
        self.path.open_file().map_err(Into::into)
    }

    #[must_use]
    /// Get the path as a [`str`]
    pub fn as_str(&self) -> &str {
        self.path.as_str()
    }

    #[must_use]
    /// Get the parent of the path
    pub fn parent(&self) -> Self {
        Self {
            path: self.path.parent(),
            workspace: self.workspace.clone(),
        }
    }

    /// Locate a path in the workspace
    ///
    /// Checks in order:
    /// - Relative to the current path, or absolute if the path starts with `/`
    /// - In the scanned pointers (prefix files)
    /// - In the include path
    ///
    /// # Errors
    /// [`Error::Vfs`] if the path could not be located
    pub fn locate(&self, path: &str) -> Result<Option<Self>, Error> {
        let path = path.replace('\\', "/");
        if path.starts_with('/') {
            if self.workspace.vfs.join(&path)?.exists()? {
                trace!("Located with absolute path: {:?}", path);
                return Ok(Some(Self {
                    path: self.workspace.vfs.join(path)?,
                    workspace: self.workspace.clone(),
                }));
            }
            if let Some((base, root)) = self
                .workspace
                .pointers
                .iter()
                .find(|(p, _)| path.starts_with(&format!("{p}/")))
            {
                let path = root.join(
                    path.strip_prefix(base)
                        .unwrap_or(&path)
                        .strip_prefix('/')
                        .unwrap_or(&path),
                )?;
                if path.exists()? {
                    trace!("Located with prefix pointer: {:?}", path);
                    return Ok(Some(Self {
                        path,
                        workspace: self.workspace.clone(),
                    }));
                }
            }
        }
        let path = self.path.parent().join(path)?;
        if path.exists()? {
            trace!("Located with parent: {:?}", path);
            Ok(Some(Self {
                path,
                workspace: self.workspace.clone(),
            }))
        } else {
            Ok(None)
        }
    }

    #[must_use]
    /// All the of addons in the workspace
    pub fn addons(&self) -> &[VfsPath] {
        &self.workspace.addons
    }

    #[must_use]
    /// All the of missions in the workspace
    pub fn missions(&self) -> &[VfsPath] {
        &self.workspace.missions
    }

    /// Walk the workspace
    ///
    /// # Errors
    /// [`Error::Vfs`] if the workspace could not be walked
    #[allow(clippy::missing_panics_doc)]
    pub fn walk_dir(&self) -> Result<Vec<Self>, Error> {
        Ok(self
            .path
            .walk_dir()?
            .filter(std::result::Result::is_ok)
            .map(move |p| Self {
                path: p.expect("filtered"),
                workspace: self.workspace.clone(),
            })
            .collect())
    }

    /// Return the metadata for the path
    ///
    /// # Errors
    /// [`Error::Vfs`] if the metadata could not be read
    pub fn metadata(&self) -> Result<vfs::VfsMetadata, Error> {
        self.path.metadata().map_err(Into::into)
    }

    #[must_use]
    /// Retruns the file name of the path
    pub fn filename(&self) -> String {
        self.path.filename()
    }

    #[must_use]
    /// Returns the extension of the path
    pub fn extension(&self) -> Option<String> {
        self.path.extension()
    }
}

impl std::hash::Hash for WorkspacePath {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.path.as_str().hash(state);
    }
}

impl std::fmt::Display for WorkspacePath {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        self.path.as_str().fmt(f)
    }
}

impl std::fmt::Debug for WorkspacePath {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        self.path.as_str().fmt(f)
    }
}
