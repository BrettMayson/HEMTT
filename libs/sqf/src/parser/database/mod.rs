//! Allows customization of the commands list at runtime in order to facilitate forwards-compatibility.

use std::collections::HashSet;

use arma3_wiki::{
    model::{Call, Version},
    Wiki,
};
use hemtt_workspace::WorkspacePath;
use tracing::{error, trace, warn};

use crate::Error;

/// The list of commands that are valid nular command constants for the compiler.
pub const NULAR_COMMANDS_CONSTANTS: &[&str] = &[
    // NOTE: `netobjnull` is not included because it's broken
    // TODO: find out if there are more commands valid as constants
    "nil",
    "confignull",
    "controlnull",
    "diaryrecordnull",
    "displaynull",
    "grpnull",
    "locationnull",
    "objnull",
    "scriptnull",
    "tasknull",
    "teammembernull",
];

/// Nular commands that are alpha-numeric and have special meaning.
///
/// Adding or removing these from a [`Database`] will do nothing, as they are handled manually by the parser.
pub const NULAR_COMMANDS_SPECIAL: &[&str] = &["true", "false"];

/// Binary commands that are alpha-numeric and have special precedence.
///
/// Adding or removing these from a [`Database`] will do nothing, as they are handled manually by the parser.
pub const BINARY_COMMANDS_SPECIAL: &[&str] = &["or", "and", "else", "max", "min", "mod", "atan2"];

/// Commands (operators) that are non-alpha-numeric or have special precedence.
pub const COMMANDS_OPERATORS: &[&str] = &[
    "!", "||", "&&", "==", "!=", ">>", ">=", "<=", ">", "<", "+", "-", "*", "/", "%", "^", ":", "#",
];

/// Contains a list of most nular, unary, and binary commands.
///
/// Instances of [`Database`] should only contain commands that are alpha-numeric, and commands that do not have special precedence.
/// Non-alpha-numeric commands and commands with special precedence are handled manually by the parser.
pub struct Database {
    nular_commands: HashSet<String>,
    unary_commands: HashSet<String>,
    binary_commands: HashSet<String>,
    wiki: Wiki,
}

impl Database {
    #[must_use]
    /// An empty database with no entries.
    pub fn empty(force_pull: bool) -> Self {
        Self {
            nular_commands: HashSet::new(),
            unary_commands: HashSet::new(),
            binary_commands: HashSet::new(),
            wiki: load_wiki(force_pull),
        }
    }

    #[must_use]
    pub fn a3(force_pull: bool) -> Self {
        let mut nular_commands = HashSet::new();
        let mut unary_commands = HashSet::new();
        let mut binary_commands = HashSet::new();

        let wiki = load_wiki(force_pull);

        for command in wiki.commands().raw().values() {
            for syntax in command.syntax() {
                match syntax.call() {
                    Call::Nular => {
                        nular_commands.insert(command.name().to_ascii_lowercase());
                    }
                    Call::Unary(_) => {
                        unary_commands.insert(command.name().to_ascii_lowercase());
                    }
                    Call::Binary(_, _) => {
                        binary_commands.insert(command.name().to_ascii_lowercase());
                    }
                }
            }
        }

        for &command in NULAR_COMMANDS_SPECIAL {
            nular_commands.remove(command);
        }
        for &command in BINARY_COMMANDS_SPECIAL {
            binary_commands.remove(command);
        }

        Self {
            nular_commands,
            unary_commands,
            binary_commands,
            wiki,
        }
    }

    /// Creates a new database with the default commands and custom commands from the workspace.
    ///
    /// # Errors
    /// [`Error::CustomCommandError`] if a custom command could not be parsed.
    ///
    /// # Panics
    /// If the custom commands directory could not be read.
    pub fn a3_with_workspace(workspace: &WorkspacePath, force_pull: bool) -> Result<Self, Error> {
        let mut database = Self::a3(force_pull);
        let custom_root = workspace.join("/.hemtt/commands");
        if let Ok(custom_root) = custom_root {
            if custom_root.exists().unwrap_or(false) {
                for entry in custom_root
                    .read_dir()
                    .expect("failed to read custom commands dir")
                {
                    if !entry.is_file().unwrap_or(false) {
                        continue;
                    }
                    let content = entry
                        .read_to_string()
                        .expect("failed to read custom command file");
                    let command = database
                        .wiki
                        .add_custom_command_parse(&content)
                        .map_err(Error::CustomCommandError)?;
                    for syntax in command.syntax() {
                        match syntax.call() {
                            Call::Nular => {
                                database
                                    .nular_commands
                                    .insert(command.name().to_ascii_lowercase());
                            }
                            Call::Unary(_) => {
                                database
                                    .unary_commands
                                    .insert(command.name().to_ascii_lowercase());
                            }
                            Call::Binary(_, _) => {
                                database
                                    .binary_commands
                                    .insert(command.name().to_ascii_lowercase());
                            }
                        }
                    }
                }
            }
        }
        Ok(database)
    }

    pub fn add_nular_command(&mut self, command: &str) {
        if is_valid_name(command) && !is_in(NULAR_COMMANDS_SPECIAL, command) {
            self.nular_commands.insert(command.to_ascii_lowercase());
        };
    }

    pub fn add_unary_command(&mut self, command: &str) {
        if is_valid_name(command) {
            self.unary_commands.insert(command.to_ascii_lowercase());
        };
    }

    pub fn add_binary_command(&mut self, command: &str) {
        if is_valid_name(command) && !is_in(BINARY_COMMANDS_SPECIAL, command) {
            self.binary_commands.insert(command.to_ascii_lowercase());
        };
    }

    #[must_use]
    pub fn has_nular_command(&self, command: &str) -> bool {
        self.nular_commands.contains(&command.to_ascii_lowercase())
    }

    #[must_use]
    pub fn has_unary_command(&self, command: &str) -> bool {
        self.unary_commands.contains(&command.to_ascii_lowercase())
    }

    #[must_use]
    pub fn has_binary_command(&self, command: &str) -> bool {
        self.binary_commands.contains(&command.to_ascii_lowercase())
    }

    #[must_use]
    pub fn has_command(&self, command: &str) -> bool {
        let command = command.to_ascii_lowercase();
        self.nular_commands.contains(&command)
            || self.unary_commands.contains(&command)
            || self.binary_commands.contains(&command)
    }

    #[must_use]
    pub const fn nular_commands(&self) -> &HashSet<String> {
        &self.nular_commands
    }

    #[must_use]
    pub const fn unary_commands(&self) -> &HashSet<String> {
        &self.unary_commands
    }

    #[must_use]
    pub const fn binary_commands(&self) -> &HashSet<String> {
        &self.binary_commands
    }

    #[must_use]
    pub const fn wiki(&self) -> &Wiki {
        &self.wiki
    }

    #[must_use]
    pub fn command_version(&self, command: &str) -> Option<&Version> {
        self.wiki
            .commands()
            .get(command)
            .and_then(|c| c.since().arma_3())
    }
}

#[must_use]
/// Whether or not this command is valid as a nular constant.
///
/// The given command must be lowercase.
pub fn is_constant_command(command: &str) -> bool {
    is_in(NULAR_COMMANDS_CONSTANTS, command)
}

#[must_use]
/// Whether or not this command is special (has special meaning or precedence).
///
/// The given command must be lowercase.
pub fn is_special_command(command: &str) -> bool {
    is_in(NULAR_COMMANDS_SPECIAL, command) || is_in(BINARY_COMMANDS_SPECIAL, command)
}

#[must_use]
/// Whether or not this command is an operator.
pub fn is_operator_command(command: &str) -> bool {
    is_in(COMMANDS_OPERATORS, command)
}

#[must_use]
/// Whether or not this command is alpha-numeric or a valid command name.
pub fn is_valid_name(command: &str) -> bool {
    command
        .chars()
        .all(|ch| ch.is_ascii_alphanumeric() || ch == '_')
}

#[must_use]
/// Whether or not this command is alpha-numeric, or an operator command (passes [`is_operator_command`]).
pub fn is_valid_command(command: &str) -> bool {
    is_valid_name(command) || is_operator_command(command)
}

fn is_in(list: &[&str], item: &str) -> bool {
    list.iter().any(|i| i.eq_ignore_ascii_case(item))
}

fn load_wiki(force_pull: bool) -> Wiki {
    Wiki::load_git(force_pull).unwrap_or_else(|e| {
        trace!(?e, "failed to load arma 3 wiki from git: {}", e);
        if force_pull {
            error!("Failed to update Arma 3 wiki from remote");
        } else {
            warn!("Failed to load Arma 3 wiki from git, falling back to bundled version");
        }
        Wiki::load_dist()
    })
}
