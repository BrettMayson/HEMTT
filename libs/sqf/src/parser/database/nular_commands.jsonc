// Nular commands (except for special characters)

[
    "acctime",
    "activatedaddons",
    "agents",
    "airdensitycurvertd",
    "all3denentities",
    "allactivetitleeffects",
    "alladdonsinfo",
    "allairports",
    "allcurators",
    "allcutlayers",
    "alldead",
    "alldeadmen",
    "alldisplays",
    "allenv3dsoundsources",
    "allgroups",
    "allmapmarkers",
    "allmines",
    "allplayers",
    "allsites",
    "allunits",
    "allunitsuav",
    "allusers",
    "ambienttemperature",
    "apertureparams",
    "armorypoints",
    "benchmark",
    "blufor",
    "break",
    "briefingname",
    "buldozer_isenabledroaddiag",
    "buldozer_reloadopermap",
    "cadetmode",
    "cameraon",
    "cameraview",
    "campaignconfigfile",
    "cansuspend",
    "cheatsenabled",
    "civilian",
    "clearforcesrtd",
    "clearitempool",
    "clearmagazinepool",
    "clearradio",
    "clearweaponpool",
    "clientowner",
    "commandingmenu",
    "configfile",
    "confignull",
    "continue",
    "controlnull",
    "copyfromclipboard",
    "createhashmap",
    "curatorcamera",
    "curatormouseover",
    "curatorselected",
    "current3denoperation",
    "currentchannel",
    "currentnamespace",
    "cursorobject",
    "cursortarget",
    "customwaypointposition",
    "date",
    "daytime",
    "diag_activemissionfsms",
    "diag_activescripts",
    "diag_activesqfscripts",
    "diag_activesqsscripts",
    "diag_allmissioneventhandlers",
    "diag_deltatime",
    "diag_fps",
    "diag_fpsmin",
    "diag_frameno",
    "diag_scope",
    "diag_stacktrace",
    "diag_ticktime",
    "dialog",
    "diaryrecordnull",
    "didjip",
    "difficulty",
    "difficultyenabledrtd",
    "disabledebriefingstats",
    "disableserialization",
    "displaynull",
    "distributionregion",
    "dynamicsimulationsystemenabled",
    "east",
    "enableenddialog",
    "endl",
    "endloadingscreen",
    "environmentenabled",
    "environmentvolume",
    "estimatedendservertime",
    "exit",
    "false",
    "finishmissioninit",
    "fog",
    "fogforecast",
    "fogparams",
    "forcedmap",
    "forceend",
    "forceweatherchange",
    "freelook",
    "get3dencamera",
    "get3deniconsvisible",
    "get3denlinesvisible",
    "get3denmouseover",
    "getartillerycomputersettings",
    "getaudiooptionvolumes",
    "getcalculateplayervisibilitybyfriendly",
    "getclientstate",
    "getclientstatenumber",
    "getcursorobjectparams",
    "getdlcassetsusage",
    "getelevationoffset",
    "getlighting",
    "getloadedmodsinfo",
    "getmissiondlcs",
    "getmissionlayers",
    "getmouseposition",
    "getmusicplayedtime",
    "getobjectviewdistance",
    "getpipviewdistance",
    "getremotesensorsdisabled",
    "getresolution",
    "getshadowdistance",
    "getsteamfriendsservers",
    "getsubtitleoptions",
    "getterraingrid",
    "getterraininfo",
    "gettiparameters",
    "gettotaldlcusagetime",
    "getvideooptions",
    "groupiconselectable",
    "groupiconsvisible",
    "grpnull",
    "gusts",
    "halt",
    "hasinterface",
    "hcshownbar",
    "hudmovementlevels",
    "humidity",
    "independent",
    "initambientlife",
    "is3den",
    "is3denmultiplayer",
    "is3denpreview",
    "isactionmenuvisible",
    "isautotest",
    "isdedicated",
    "isfilepatchingenabled",
    "isgamefocused",
    "isgamepaused",
    "isinstructorfigureenabled",
    "ismissionprofilenamespaceloaded",
    "ismultiplayer",
    "ismultiplayersolo",
    "ispipenabled",
    "isremoteexecuted",
    "isremoteexecutedjip",
    "issaving",
    "isserver",
    "issteammission",
    "issteamoverlayenabled",
    "isstreamfriendlyuienabled",
    "isstressdamageenabled",
    "istuthintsenabled",
    "isuicontext",
    "language",
    "librarycredits",
    "librarydisclaimers",
    "lightnings",
    "linebreak",
    "loadgame",
    "localnamespace",
    "locationnull",
    "logentities",
    "mapanimclear",
    "mapanimcommit",
    "mapanimdone",
    "markasfinishedonsteam",
    "missionconfigfile",
    "missiondifficulty",
    "missionend",
    "missionname",
    "missionnamesource",
    "missionnamespace",
    "missionprofilenamespace",
    "missionstart",
    "missionversion",
    "moonintensity",
    "musicvolume",
    "netobjnull",
    "nextweatherchange",
    "nil",
    "objnull",
    "opencuratorinterface",
    "opfor",
    "overcast",
    "overcastforecast",
    "parsingnamespace",
    "particlesquality",
    "pi",
    "pixelgrid",
    "pixelgridbase",
    "pixelgridnouiscale",
    "pixelh",
    "pixelw",
    "playableunits",
    "player",
    "playerrespawntime",
    "playerside",
    "productversion",
    "profilename",
    "profilenamespace",
    "profilenamesteam",
    "radioenabled",
    "radiovolume",
    "rain",
    "rainbow",
    "rainparams",
    "remoteexecutedowner",
    "resetcamshake",
    "resistance",
    "reversedmousey",
    "runinitscript",
    "safezoneh",
    "safezonew",
    "safezonewabs",
    "safezonex",
    "safezonexabs",
    "safezoney",
    "savegame",
    "savejoysticks",
    "savemissionprofilenamespace",
    "saveprofilenamespace",
    "savingenabled",
    "scriptnull",
    "selectnoplayer",
    "sentencesenabled",
    "servername",
    "servernamespace",
    "servertime",
    "shownartillerycomputer",
    "shownchat",
    "showncompass",
    "showncuratorcompass",
    "showngps",
    "shownhud",
    "shownmap",
    "shownpad",
    "shownradio",
    "shownscoretable",
    "shownsubtitles",
    "shownuavfeed",
    "shownwarrant",
    "shownwatch",
    "sideambientlife",
    "sideempty",
    "sideenemy",
    "sidefriendly",
    "sidelogic",
    "sideunknown",
    "simulweathersync",
    "slingloadassistantshown",
    "soundvolume",
    "speechvolume",
    "sunormoon",
    "switchableunits",
    "systemofunits",
    "systemtime",
    "systemtimeutc",
    "tasknull",
    "teammembernull",
    "teams",
    "teamswitch",
    "teamswitchenabled",
    "time",
    "timemultiplier",
    "true",
    "uinamespace",
    "userinputdisabled",
    "vehicles",
    "viewdistance",
    "visiblecompass",
    "visiblegps",
    "visiblemap",
    "visiblescoretable",
    "visiblewatch",
    "waves",
    "west",
    "wind",
    "winddir",
    "windrtd",
    "windstr",
    "worldname",
    "worldsize"
  ]
