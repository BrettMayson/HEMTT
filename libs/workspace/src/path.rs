use std::{hash::Hasher, sync::Arc};

use hemtt_common::strip::StripInsensitive;
use vfs::{SeekAndWrite, VfsPath};

use super::{Error, LayerType, Workspace};

#[derive(Clone, PartialEq, Eq)]
pub struct WorkspacePathData {
    pub(crate) path: VfsPath,
    pub(crate) workspace: Arc<Workspace>,
}

#[allow(clippy::module_name_repetitions)]
#[derive(Clone, PartialEq, Eq)]
/// A path in a workspace
pub struct WorkspacePath {
    pub(crate) data: Arc<WorkspacePathData>,
}

impl WorkspacePath {
    #[must_use]
    /// Returns the underlying [`VfsPath`]
    pub fn vfs(&self) -> &VfsPath {
        &self.data.path
    }

    #[must_use]
    /// Returns the workspace
    pub fn workspace(&self) -> &Workspace {
        &self.data.workspace
    }

    #[must_use]
    /// Is the file from an include path
    pub fn is_include(&self) -> bool {
        self.data
            .workspace
            .layers
            .iter()
            .filter(|(_, t)| *t == LayerType::Include)
            .any(|(p, _)| {
                p.join(self.data.path.as_str())
                    .and_then(|p| p.exists())
                    .unwrap_or(false)
            })
    }

    /// join a path to the workspace path
    ///
    /// # Errors
    /// [`Error::Vfs`] if the path could not be joined
    pub fn join(&self, path: impl AsRef<str>) -> Result<Self, Error> {
        let path = self.data.path.join(path)?;
        Ok(Self {
            data: Arc::new(WorkspacePathData {
                path,
                workspace: self.data.workspace.clone(),
            }),
        })
    }

    /// Create a file in the workspace
    ///
    /// # Errors
    /// [`Error::Vfs`] if the file could not be created
    pub fn create_file(&self) -> Result<Box<dyn SeekAndWrite + Send>, Error> {
        self.data.path.create_file().map_err(Into::into)
    }

    /// Create a directory in the workspace
    ///
    /// # Errors
    /// [`Error::Vfs`] if the directory could not be created
    pub fn create_dir(&self) -> Result<(), Error> {
        self.data.path.create_dir()?;
        Ok(())
    }

    /// Check if the path exists
    ///
    /// # Errors
    /// [`Error::Vfs`] if the path could not be checked
    pub fn exists(&self) -> Result<bool, Error> {
        self.data.path.exists().map_err(Into::into)
    }

    /// Check if the path is a file
    ///
    /// # Errors
    /// [`Error::Vfs`] if the path could not be checked
    pub fn is_file(&self) -> Result<bool, Error> {
        self.data.path.is_file().map_err(Into::into)
    }

    /// Check if the path is a directory
    ///
    /// # Errors
    /// [`Error::Vfs`] if the path could not be checked
    pub fn is_dir(&self) -> Result<bool, Error> {
        self.data.path.is_dir().map_err(Into::into)
    }

    /// Read the path to a string
    ///
    /// # Errors
    /// [`Error::Vfs`] if the path could not be read
    pub fn read_to_string(&self) -> Result<String, Error> {
        self.data
            .path
            .read_to_string()
            .map(|s| s.replace('\r', ""))
            .map_err(Into::into)
    }

    /// Open the file for reading
    ///
    /// # Errors
    /// [`Error::Vfs`] if the file could not be opened
    pub fn open_file(&self) -> Result<Box<dyn vfs::SeekAndRead + Send>, Error> {
        self.data.path.open_file().map_err(Into::into)
    }

    #[must_use]
    /// Get the path as a [`str`]
    pub fn as_str(&self) -> &str {
        self.data.path.as_str()
    }

    #[must_use]
    /// Get the parent of the path
    pub fn parent(&self) -> Self {
        Self {
            data: Arc::new(WorkspacePathData {
                path: self.data.path.parent(),
                workspace: self.data.workspace.clone(),
            }),
        }
    }

    /// Change the extension of the path
    ///
    /// # Errors
    /// [`Error::Vfs`] if the path could not be changed
    pub fn with_extension(&self, ext: &str) -> Result<Self, Error> {
        Ok(Self {
            data: Arc::new(WorkspacePathData {
                path: {
                    let current = self
                        .data
                        .path
                        .filename()
                        .chars()
                        .rev()
                        .collect::<String>()
                        .split_once('.')
                        .map_or(
                            self.data.path.filename().as_str().chars().rev().collect(),
                            |(_, s)| s.to_string(),
                        )
                        .chars()
                        .rev()
                        .collect::<String>();
                    self.data.path.parent().join(format!("{current}.{ext}"))?
                },
                workspace: self.data.workspace.clone(),
            }),
        })
    }

    /// Locate a path in the workspace
    ///
    /// Checks in order:
    /// - A3 P drive, if allowed and path starts with `/a3/`
    /// - Relative to the current path, or absolute if the path starts with `/`
    /// - In the scanned pointers (prefix files)
    /// - In the include path
    ///
    /// # Errors
    /// [`Error::Vfs`] if the path could not be located
    pub fn locate(&self, path: &str) -> Result<Option<Self>, Error> {
        let path = path.replace('\\', "/");
        let path_lower = path.to_lowercase();
        if path_lower.starts_with("/a3/") {
            if let Some(pdrive) = &self.workspace().pdrive {
                if let Some(pdrive_path) = pdrive.path_to(&path) {
                    return Ok(Some(Self {
                        data: Arc::new(WorkspacePathData {
                            path: pdrive_path,
                            workspace: self.data.workspace.clone(),
                        }),
                    }));
                }
            }
        }
        if path.starts_with('/') {
            if self.data.workspace.vfs.join(&path)?.exists()? {
                return Ok(Some(Self {
                    data: Arc::new(WorkspacePathData {
                        path: self.data.workspace.vfs.join(path)?,
                        workspace: self.data.workspace.clone(),
                    }),
                }));
            }
            if let Some((base, root)) = self
                .data
                .workspace
                .pointers
                .iter()
                .find(|(p, _)| path_lower.starts_with(&format!("{}/", p.to_lowercase())))
            {
                // Windows needs case insensitivity because p3ds are a
                // disaster. On Linux we'll be more strict to avoid
                // pain and suffering.
                let path = if cfg!(windows) {
                    root.join(
                        path_lower
                            .strip_prefix(&base.to_lowercase())
                            .unwrap_or(&path)
                            .strip_prefix('/')
                            .unwrap_or(&path),
                    )?
                } else {
                    root.join(
                        path.strip_prefix_insensitive(base)
                            .unwrap_or(&path)
                            .strip_prefix('/')
                            .unwrap_or(&path),
                    )?
                };
                if path.exists()? {
                    return Ok(Some(Self {
                        data: Arc::new(WorkspacePathData {
                            path,
                            workspace: self.data.workspace.clone(),
                        }),
                    }));
                }
            }
        }
        let path = self.data.path.parent().join(path)?;
        if path.exists()? {
            Ok(Some(Self {
                data: Arc::new(WorkspacePathData {
                    path,
                    workspace: self.data.workspace.clone(),
                }),
            }))
        } else {
            Ok(None)
        }
    }

    #[must_use]
    /// All the of addons in the workspace
    pub fn addons(&self) -> &[VfsPath] {
        &self.data.workspace.addons
    }

    #[must_use]
    /// All the of missions in the workspace
    pub fn missions(&self) -> &[VfsPath] {
        &self.data.workspace.missions
    }

    /// Walk the workspace
    ///
    /// # Errors
    /// [`Error::Vfs`] if the workspace could not be walked
    #[allow(clippy::missing_panics_doc)]
    pub fn walk_dir(&self) -> Result<Vec<Self>, Error> {
        Ok(self
            .data
            .path
            .walk_dir()?
            .filter(std::result::Result::is_ok)
            .map(|p| Self {
                data: Arc::new(WorkspacePathData {
                    path: p.expect("filtered"),
                    workspace: self.data.workspace.clone(),
                }),
            })
            .collect())
    }

    /// Read the files in a directory
    ///
    /// # Errors
    /// [`Error::Vfs`] if the directory could not be read
    pub fn read_dir(&self) -> Result<Vec<Self>, Error> {
        Ok(self
            .data
            .path
            .read_dir()?
            .map(|p| Self {
                data: Arc::new(WorkspacePathData {
                    path: p,
                    workspace: self.data.workspace.clone(),
                }),
            })
            .collect())
    }

    /// Return the metadata for the path
    ///
    /// # Errors
    /// [`Error::Vfs`] if the metadata could not be read
    pub fn metadata(&self) -> Result<vfs::VfsMetadata, Error> {
        self.data.path.metadata().map_err(Into::into)
    }

    #[must_use]
    /// Retruns the file name of the path
    pub fn filename(&self) -> String {
        self.data.path.filename()
    }

    #[must_use]
    /// Returns the extension of the path
    pub fn extension(&self) -> Option<String> {
        self.data.path.extension()
    }
}

impl std::hash::Hash for WorkspacePath {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.data.path.as_str().hash(state);
    }
}

impl std::fmt::Display for WorkspacePath {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        self.data.path.as_str().fmt(f)
    }
}

impl std::fmt::Debug for WorkspacePath {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        self.data.path.as_str().fmt(f)
    }
}
