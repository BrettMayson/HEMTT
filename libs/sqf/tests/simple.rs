use std::path::PathBuf;

use hemtt_preprocessor::Processor;
use hemtt_sqf::parser::database::Database;

const ROOT: &str = "tests/simple/";

macro_rules! simple {
    ($dir:ident) => {
        paste::paste! {
            #[test]
            fn [<simple_ $dir>]() {
                simple(stringify!($dir));
            }
        }
    };
}

fn simple(file: &str) {
    let workspace = hemtt_common::workspace::Workspace::builder()
        .physical(&PathBuf::from(ROOT))
        .finish(None)
        .unwrap();
    let source = workspace.join(format!("{file}.sqf")).unwrap();
    let processed = Processor::run(&source).unwrap();
    std::fs::write(format!("tests/simple/{file}.sqfp"), processed.as_str()).unwrap();
    let parsed = match hemtt_sqf::parser::run(&Database::default(), &processed) {
        Ok(sqf) => sqf,
        Err(hemtt_sqf::parser::ParserError::ParsingError(e)) => {
            for error in e {
                println!("{}", error.report().unwrap());
            }
            panic!("failed to parse");
        }
        Err(e) => panic!("{e:?}"),
    };
    assert_ne!(parsed.content().len(), 0);
    let mut buffer = Vec::new();
    parsed.compile_to_writer(&processed, &mut buffer).unwrap();
    std::fs::write(format!("tests/simple/{file}.sqfc"), buffer).unwrap();
    std::fs::write(
        format!("tests/simple/{file}.sqfast"),
        format!("{:#?}", parsed.content()),
    )
    .unwrap();
}

simple!(format_font);
simple!(dev);
simple!(eventhandler);
simple!(foreach);
simple!(get_visibility);
simple!(hash_select);
simple!(hello);
simple!(include);
simple!(oneline);
simple!(semicolons);
