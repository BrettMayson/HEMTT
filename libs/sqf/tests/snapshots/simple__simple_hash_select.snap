---
source: libs/sqf/tests/simple.rs
expression: ast
---
[
    Expression(
        BinaryCommand(
            Select,
            Variable(
                "_explosive",
                0..10,
            ),
            Number(
                FloatOrd(
                    0.0,
                ),
                13..14,
            ),
            11..12,
        ),
        0..14,
    ),
]
